### PR TITLE
[CLI] Add include other config files feature when parsing config file to LightningCLI

### LINKED ISSUE #11221: [CLI] Add include other config files feature when parsing config file to LightningCLI
## 🚀 Feature

Add ability to include other config file when we use LightningCLI and config arg.

### Motivation

If we want to train and test many models with same/many datasets, we may want to split the config file into many parts like trainer part, model part and dataset part. And then, the command to run our model will look like this: 
```bash
python tools/cil.py --config configs/default_runtime.yaml --config configs/datasets/xxx.yaml --config configs/models/xxx.yaml [--config some/thing/else.yaml] [--some.other thing]
```
LightningCLI allow us to include many config files in the same time, but the command will be too long to write. Therefore, we can write another config file looks like:

configs/runs/xxx.yaml:
```yaml
__base__:
    - ../default_runtime.yaml
    - ../datasets/xxx.yaml
    - ../models/xxx.yaml

some:
    other: thing
```

Then, the command to run our model will look like:
```bash
python tools/cil.py --config configs/runs/xxx.yaml
```
It will be easier to write and remeber.

### Alternatives

in pytorch_lightning/utilities/cli.py [line 125](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/utilities/cli.py#:~:text=%22%2D%2Dconfig%22%2C%20action%3D-,ActionConfigFile,-%2C%20help%3D%22Path%20to)

class LightningArgumentParser add argument --config with action ActionConfigFile from jsonargparse package, and we need override this action.

```python
def deep_update(source, override):
    """
    Update a nested dictionary or similar mapping.
    Modify ``source`` in place.
    """
    if isinstance(source, Dict) and isinstance(override, Dict):
        if '__delete__' in override:
            delete_keys = override.pop('__delete__')
            if isinstance(delete_keys, str):
                delete_keys = [delete_keys]

            if isinstance(delete_keys, list):
                for k in delete_keys:
                    if k in source:
                        source.pop(k)
            elif delete_keys:
                return override
        for key, value in override.items():
            if isinstance(value, Dict) and key in source:
                source[key] = deep_update(source[key], value)
            else:
                source[key] = override[key]
        return source
    else:
        return override


def parse_dict_config(parser, cfg_file, cfg_path = None, **kwargs):
    if '__base__' in cfg_file:
        sub_cfg_paths = cfg_file.pop('__base__')
        if sub_cfg_paths is not None:
            if not isinstance(sub_cfg_paths, list):
                sub_cfg_paths = [sub_cfg_paths]
            sub_cfg_paths = [os.path.join(os.path.dirname(cfg_path), sub_cfg_path) if not sub_cfg_path.startswith(
                '/') or cfg_path is None else sub_cfg_path for sub_cfg_path in sub_cfg_paths]
            sub_cfg_file = {}
            for sub_cfg_path in sub_cfg_paths:
                sub_cfg_file = deep_update(sub_cfg_file, parse_path(parser, sub_cfg_path, **kwargs).as_dict())
            cfg_file = deep_update(sub_cfg_file, cfg_file)
    if '__import__' in cfg_file:
        cfg_file.pop('__import__')

    for k, v in cfg_file.items():
        if isinstance(v, dict):
            cfg_file[k] = parse_dict_config(parser, v, cfg_path, **kwargs)
    return cfg_file


def parse_config(parser, cfg_file, cfg_path = None, **kwargs):
    return parser._apply_actions(parse_dict_config(parser, cfg_file.as_dict(), cfg_path, **kwargs))


def parse_path(parser, cfg_path, seen_cfg = None, **kwargs):
    abs_cfg_path = os.path.abspath(cfg_path)
    if seen_cfg is None:
        seen_cfg = {}
    elif abs_cfg_path in seen_cfg:
        if seen_cfg[abs_cfg_path] is None:
            raise RuntimeError('Circular reference detected in config file')
        else:
            return seen_cfg[abs_cfg_path]

    cfg_file = parser.parse_path(cfg_path, **kwargs)
    seen_cfg[abs_cfg_path] = None
    cfg_file = parse_config(parser, cfg_file, cfg_path = cfg_path, seen_cfg = seen_cfg, **kwargs)
    seen_cfg[abs_cfg_path] = cfg_file
    return cfg_file


def parse_string(parser, cfg_string, **kwargs):
    return parse_config(parser, parser.parse_string(cfg_string, **kwargs), **kwargs)


class LightningActionConfigFile(ActionConfigFile):
    @staticmethod
    def apply_config(parser, cfg, dest, value) -> None:
        with _ActionSubCommands.not_single_subcommand():
            if dest not in cfg:
                cfg[dest] = []
            kwargs = {'env': False, 'defaults': False, '_skip_check': True, '_fail_no_subcommand': False}
            try:
                cfg_path: Optional[Path] = Path(value, mode = get_config_read_mode())
            except TypeError as ex_path:
                try:
                    if isinstance(load_value(value), str):
                        raise ex_path
                    cfg_path = None
                    cfg_file = parse_string(parser, value, **kwargs)
                except (TypeError,) + get_loader_exceptions() as ex_str:
                    raise TypeError(f'Parser key "{dest}": {ex_str}') from ex_str
            else:
                cfg_file = parse_path(parser, value, **kwargs)
            cfg[dest].append(cfg_path)
            cfg.update(cfg_file)
```

### Additional context

The code of pop '__import__' from config file is for easier use of reference of yaml file. We can write the yaml file more clearly using it. An example is when we use the dataset, we may want to define the transforms in the config files, too. And by this little trick, we can make the transforms more clearly like:

```yaml
__import__:
   import_transform_2: &import_transform_2
         some: thing
    import_train_transforms: &import_train_transforms
        - transform_1
        - *import_transform_2
        - transform_3
    import_test_transforms: &import_test_transforms
        - transform_1
        - *import_transform_2
        - transform_3

data:
    train_transforms: *import_train_transforms
    test_transforms: *import_train_transforms
```

We can do it without this feature in this little case too, but the config file may be ugly and misleading.
______________________________________________________________________

#### If you enjoy Lightning, check out our other projects! ⚡

- [**Metrics**](https://github.com/PyTorchLightning/metrics): Machine learning metrics for distributed, scalable PyTorch applications.

- [**Lite**](https://pytorch-lightning.readthedocs.io/en/latest/starter/lightning_lite.html): enables pure PyTorch users to scale their existing code on any kind of device while retaining full control over their own loops and optimization logic.

- [**Flash**](https://github.com/PyTorchLightning/lightning-flash): The fastest way to get a Lightning baseline! A collection of tasks for fast prototyping, baselining, fine-tuning, and solving problems with deep learning.

- [**Bolts**](https://github.com/PyTorchLightning/lightning-bolts): Pretrained SOTA Deep Learning models, callbacks, and more for research and production with PyTorch Lightning and PyTorch.

- [**Lightning Transformers**](https://github.com/PyTorchLightning/lightning-transformers): Flexible interface for high-performance research using SOTA Transformers leveraging Pytorch Lightning, Transformers, and Hydra.

cc @borda @carmocca @mauvilsa

### PR DESCRIPTION
**Pull request by shenmishajing**
## What does this PR do?

Fixes #11221

### Does your PR introduce any breaking changes? If yes, please list them.
None

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [ ] Did you make sure to **update the documentation** with your changes? (if necessary)
- [ ] Did you write any **new necessary tests**? (not for typos and docs)
- [ ] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [ ] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

**Comment by carmocca** [comment]
As per the motivation in https://github.com/PyTorchLightning/pytorch-lightning/issues/11221#issuecomment-1003401375, closing this pull request.
